psh/date/date.c

```
@@ -0,0 +1,153 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+char *wdaylookup[] = { "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat" };
+
+char *monlookup[] = { "Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec" };
+
+enum mode { SKIP,
+	HELP,
+	PRINT,
+	WRITE };
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf("Usage: %s [option] [epoch]\n", prog);
+	printf("  -h:  shows this help message\n");
+	printf("  -s:  set system time described by EPOCH (POSIX time format)\n");
+}
+
+
+static int psh_date_print(void)
+{
+	time_t t;
+	struct tm *tinfo;
+
+	time(&t);
+	tinfo = localtime(&t);
+
+	printf("%s,", wdaylookup[tinfo->tm_wday]);
+	/* `struct tm` field tm_year is years since 1900, so modulo for last 2 digits */
+	printf(" %02d %s %d", tinfo->tm_mday, monlookup[tinfo->tm_mon], (tinfo->tm_year) % 100);
+	printf(" %02d:%02d:%02d", tinfo->tm_hour, tinfo->tm_min, tinfo->tm_sec);
+	printf("\n");
+
+	return 0;
+}
+
+
+static int psh_date_set(char *timestring)
+{
+	struct timeval tv = { .tv_sec = 0, .tv_usec = 0 };
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	if (strlen(timestring) == 1 && timestring[0] == '0') {
+		tv.tv_sec = 0;
+	}
+	else {
+		tv.tv_sec = strtoull(timestring, NULL, 10);
+		if (tv.tv_sec == 0) {
+			fprintf(stderr, "Uncrecognized argument \"%s\"!\n", timestring);
+			return 1;
+		}
+	}
+
+	return settimeofday(&tv, &tz);
+}
+
+
+static int psh_date(int argc, char **argv)
+{
+	char c;
+	int status = 1;
+	char *timestring = NULL;
+	enum mode m = (argc == 1) ? PRINT : SKIP;
+
+	while ((c = getopt(argc, argv, "hs:")) != -1) {
+		switch (c) {
+			case 'h':
+				m = HELP;
+				break;
+
+			case 's':
+				m = WRITE;
+				timestring = optarg;
+				break;
+
+			case ':':
```

no need to handle these suboptions here, just fail on default - getopt will write error regarding missing paramter

---

Changed! Thanks!